README.md
# Power Commit

Introducing the ultimate productivity booster for developers and programmers: the Visual Studio Code Extension to quickly commit to Git!

Say goodbye to the tedium of manual commits and embrace lightning-fast workflows with just a single hotkey.

This game-changing extension empowers you to submit changes instantly, shaving off countless precious seconds and boosting your efficiency like never before. With thousands of iterations, it's the secret weapon that transforms you into a speed demon of version control.

Get ready to supercharge your development process and experience the exhilaration of rapid commits with this must-have extension!

## [Extension Home Page](https://marketplace.visualstudio.com/items?itemName=Eshnek.power-commit&ssr=false)

## Usage

Bind a keyboard shortcut to the `powercommit.powerCommit` command.

Alternatively, manually activate the extension by searching for it in the context menu.

**Warning: Generally incompatible with team work, unless people are okay with timestamp commits.**

**Warning: May have unintended consequences. Be sure to read below to understand exactly what the extension does.**

## Details

When invoked, the extension will:
1. Report a notification to the user via the status bar.
1. Determine the root path of the currently selected editor.
1. Save all files.
1. Execute `git add .`, staging all files.
1. Execute `git commit -m "..."`, committing a timestamp message.
1. Execute `git pull`
1. Execute `git push`
1. Report a notification to the user via the status bar.

All commands are run in the determined editor root folder.

The timestamp typically looks like this: `6/21/2023 14:28 +`

Nothing happens when there are no changes - Git simply rejects the commands.


CHANGELOG.md
# Change Log

All notable changes to the "power-commit" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Released]

- Initial release

## [1.0.5] - 6/21/2023

### Changes

- No punctuation in notification
- Improve README with warnings

## [1.0.6] - 6/21/2023

### Changes

- Rename project from `auto-commit` to `power-commit`

## [1.0.7] - 6/21/2023

### Changes

- Project display name has a space in it
- Update extension description

## [1.1.0] - 6/26/2023

### Changes

- Reporting now occurs in the status bar, as opposed to using a notification. This is less invasive.

## [1.1.1] - 6/26/2023

### Fixed

- Error reporting bug causing functionality to break

## [1.1.2] - 6/27/2023

### Changes

- Added startup activation event

## [1.2.0] - 6/30/2023

- Added a new icon


src/test/suite/extension.test.ts
import * as assert from 'assert';
import * as util from '../../util';

import * as mocha from 'mocha';
const describe = mocha.describe;

suite('Extension Test Suite', () => {
    describe('util', () => {
        test('util.system', async () => {
            // (vscode.workspace.workspaceFolders as any) = [{url: {fsPath: TEST_FOLDER}}];

            // We did not open a folder
            assert.strictEqual(await util.system('pwd'), '');
        });
    });
});


src/util.ts
import * as vscode from 'vscode';

const exec = require('child_process').exec;

//-//

const STATUS_TIMEOUT: number = 5000;

const DEFAULT_REPORTER: (text: string) => void = (text: string) => {
    console.error(text);

    // We get a non-zero exit code in certain valid cases.
    // throw new Error(text);
};

//-//

let statusBar: vscode.StatusBarItem | null = null;
let statusTimeout: any | null = null;

//-//

export async function system(
    command: string,
    cwd: string | null = null,
    reporter: (text: string) => void = DEFAULT_REPORTER
): Promise<string> {
    if (cwd === null) {
        if ((cwd = getCWD()) === null) {
            return '';
        }
    }

    return new Promise((resolve) => {
        exec(command, { cwd }, makeCommandResolver(resolve, reporter));
    });
}
function makeCommandResolver(resolve: any, reporter: (text: string) => void): any {
    return (error: string, stdout: string, stderr: string) => {
        if (error || stderr) {
            reporter(error || stderr);
        }

        resolve(stdout);
    };
}
function getCWD(): string | null {
    if (vscode.workspace.workspaceFolders === undefined) {
        return null;
    }

    return `${vscode.workspace.workspaceFolders[0].uri.fsPath}`;
}

//-//

export async function sleep(ms: number) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

export function inform(text: string): void {
    console.info(text);

    showStatus(text);
}
function showStatus(text: string): void {
    if(statusBar === null) {
        createStatusBar();
    }

    statusBar!.text = text;
    statusBar!.show();

    startStatusTimer();
}
function createStatusBar(): void {
    statusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1000);
}
function startStatusTimer(): void {
    if (statusTimeout !== null) {
        clearTimeout(statusTimeout);
    }

    statusTimeout = setTimeout(() => {
        statusBar!.text = '';
        statusBar!.hide();
    }, STATUS_TIMEOUT);
}


//-//

export function getActiveDocument(): vscode.TextDocument | undefined {
    return getActiveEditor()?.document;
}
export function getActiveEditor(): vscode.TextEditor | undefined {
    return vscode.window.activeTextEditor;
}


src/autoCommit.ts
import * as vscode from 'vscode';
import * as util from './util';

//-//

const SYNC_DELAY = 0;
const AUTO_COMMIT_TEXT_START = 'Performing auto commit';
const AUTO_COMMIT_TEXT_DONE = 'Auto commit complete!';

let currentlyAutoCommitting: boolean = false;

//-//

export async function go(): Promise<void> {
    const commit = new AutoCommit();

    await commit.commit();
}

//-//

class AutoCommit {

    private topLevel: string | null = null;

    public async commit(): Promise<void> {
        if (currentlyAutoCommitting) {
            return;
        }
        currentlyAutoCommitting = true;

        await this.commit_();

        currentlyAutoCommitting = false;
    }
    private async commit_(): Promise<void> {
        await AutoCommit.pushPending();

        await this.performAutoCommit();

        await AutoCommit.popPending();
    }

    private static async pushPending(): Promise<void> {
        util.inform(AUTO_COMMIT_TEXT_START);
    }
    private static async popPending(): Promise<void> {
        util.inform(AUTO_COMMIT_TEXT_DONE);
    }

    private async performAutoCommit(): Promise<void> {
        try {
            await this.performAutoCommit_();
        } catch (e: any) {
            console.error(e);
            vscode.window.showErrorMessage(e.message);
        }
    }
    private async performAutoCommit_(): Promise<void> {
        await this.storeTopLevel();

        await this.saveAll();
        await this.stageAll();

        await this.performCommit();

        await this.syncAll();
    }
    private async storeTopLevel(): Promise<void> {
        this.topLevel = await AutoCommit.findGitRoot();
    }

    //-//

    private async saveAll(): Promise<void> {
        await vscode.commands.executeCommand('workbench.action.files.saveFiles');
    }
    private async stageAll(): Promise<void> {
        // 'git.stageAll'
        await util.system('git add .', this.topLevel);
    }

    private async performCommit(): Promise<void> {
        const message = AutoCommit.generateCommitMessage();

        await util.system(`git commit -m "${message}"`, this.topLevel);
    }

    private async syncAll(): Promise<void> {
        await util.sleep(SYNC_DELAY);

        // 'git.sync', 'git._syncAll'
        await util.system('git pull', this.topLevel);
        await util.system('git push', this.topLevel);
    }

    //-//

    private static async findGitRoot(): Promise<string | null> {
        const doc = util.getActiveDocument();
        if (!doc) {
            return null;
        }

        return await AutoCommit.findGitRoot_(doc);
    }
    private static async findGitRoot_(doc: vscode.TextDocument): Promise<string | null> {
        const currentFolder = AutoCommit.getCurrentFolder(doc);
        const topLevel = await AutoCommit.getTopLevel(currentFolder);

        return topLevel || null;
    }
    private static getCurrentFolder(doc: vscode.TextDocument): string {
        const currentFilePath = doc.uri.fsPath.trim().replace(/\\/g, '/');
        const currentFolder = currentFilePath.substring(0, currentFilePath.lastIndexOf('/'));

        AutoCommit.printFolder(currentFolder);

        return currentFolder;
    }
    private static async getTopLevel(currentFolder: string): Promise<string> {
        const topLevelResult = await util.system('git rev-parse --show-toplevel', currentFolder);
        const topLevel = topLevelResult.trim();

        AutoCommit.printTopLevel(topLevel);

        return topLevel;
    }

    //-//

    private static generateCommitMessage(): string {
        const now = new Date(Date.now());

        const displayedMinutes = AutoCommit.getDisplayedMinutes(now);

        // 6/21/2023 04:00 +
        return `${now.getMonth() + 1}/${now.getDate()}/${now.getFullYear()} ${now.getHours()}:${displayedMinutes} +`;
    }
    private static getDisplayedMinutes(now: Date): string {
        let displayedMinutes = `${now.getMinutes()}`;

        if (displayedMinutes.length < 2) {
            displayedMinutes = '0' + displayedMinutes;
        }

        return `${displayedMinutes}`;
    }

    //-//

    private static printFolder(folder: string): void {
        console.log(`Auto commit current folder: ${folder}`);
    }
    private static printTopLevel(folder: string): void {
        console.log(`Auto commit top-level folder: ${folder}`);
    }

};


src/powerCommit.ts
import * as vscode from 'vscode';
import * as util from './util';

//-//

const SYNC_DELAY = 0;
const COMMIT_TEXT_START = 'Performing power commit...';
const COMMIT_TEXT_DONE = 'Power commit complete!';

let currentlyPowerCommitting: boolean = false;

//-//

export async function go(): Promise<void> {
    const commit = new PowerCommit();

    await commit.commit();
}

//-//

class PowerCommit {

    private topLevel: string | null = null;

    public async commit(): Promise<void> {
        if (currentlyPowerCommitting) {
            return;
        }
        currentlyPowerCommitting = true;

        await this.commit_();

        currentlyPowerCommitting = false;
    }
    private async commit_(): Promise<void> {
        await PowerCommit.pushPending();

        await this.performPowerCommit();

        await PowerCommit.popPending();
    }

    private static async pushPending(): Promise<void> {
        util.inform(COMMIT_TEXT_START);
    }
    private static async popPending(): Promise<void> {
        util.inform(COMMIT_TEXT_DONE);
    }

    private async performPowerCommit(): Promise<void> {
        try {
            await this.performPowerCommit_();
        } catch (e: any) {
            console.error(e);
            vscode.window.showErrorMessage(e.message);
        }
    }
    private async performPowerCommit_(): Promise<void> {
        await this.storeTopLevel();

        await this.saveAll();
        await this.stageAll();

        await this.performCommit();

        await this.syncAll();
    }
    private async storeTopLevel(): Promise<void> {
        this.topLevel = await PowerCommit.findGitRoot();
    }

    //-//

    private async saveAll(): Promise<void> {
        await vscode.commands.executeCommand('workbench.action.files.saveFiles');
    }
    private async stageAll(): Promise<void> {
        // 'git.stageAll'
        await util.system('git add .', this.topLevel);
    }

    private async performCommit(): Promise<void> {
        const message = PowerCommit.generateCommitMessage();

        await util.system(`git commit -m "${message}"`, this.topLevel);
    }

    private async syncAll(): Promise<void> {
        await util.sleep(SYNC_DELAY);

        // 'git.sync', 'git._syncAll'
        await util.system('git pull', this.topLevel);
        await util.system('git push', this.topLevel);
    }

    //-//

    private static async findGitRoot(): Promise<string | null> {
        const doc = util.getActiveDocument();
        if (!doc) {
            return null;
        }

        return await PowerCommit.findGitRoot_(doc);
    }
    private static async findGitRoot_(doc: vscode.TextDocument): Promise<string | null> {
        const currentFolder = PowerCommit.getCurrentFolder(doc);
        const topLevel = await PowerCommit.getTopLevel(currentFolder);

        return topLevel || null;
    }
    private static getCurrentFolder(doc: vscode.TextDocument): string {
        const currentFilePath = doc.uri.fsPath.trim().replace(/\\/g, '/');
        const currentFolder = currentFilePath.substring(0, currentFilePath.lastIndexOf('/'));

        PowerCommit.printFolder(currentFolder);

        return currentFolder;
    }
    private static async getTopLevel(currentFolder: string): Promise<string> {
        const topLevelResult = await util.system('git rev-parse --show-toplevel', currentFolder);
        const topLevel = topLevelResult.trim();

        PowerCommit.printTopLevel(topLevel);

        return topLevel;
    }

    //-//

    private static generateCommitMessage(): string {
        const now = new Date(Date.now());

        const displayedMinutes = PowerCommit.getDisplayedMinutes(now);

        // 6/21/2023 04:00 +
        return `${now.getMonth() + 1}/${now.getDate()}/${now.getFullYear()} ${now.getHours()}:${displayedMinutes} +`;
    }
    private static getDisplayedMinutes(now: Date): string {
        let displayedMinutes = `${now.getMinutes()}`;

        if (displayedMinutes.length < 2) {
            displayedMinutes = '0' + displayedMinutes;
        }

        return `${displayedMinutes}`;
    }

    //-//

    private static printFolder(folder: string): void {
        console.log(`Auto commit current folder: ${folder}`);
    }
    private static printTopLevel(folder: string): void {
        console.log(`Auto commit top-level folder: ${folder}`);
    }

};
